src/metaxy/graph/status.py

```diff
@@ -0,0 +1,218 @@
+            elif self.needs_update:
+                status_icon = "[yellow]⚠[/yellow]"
+                status_text = "needs update"
+            else:
+                status_icon = "[green]✓[/green]"
+                status_text = "up-to-date"
+
+        return (
+            f"{status_icon} {self.feature_key.to_string()} "
+            f"(rows: {self.row_count}, added: {self.added_count}, "
+            f"changed: {self.changed_count}) — {status_text}"
+        )
+
+
+class FeatureMetadataStatusWithIncrement(NamedTuple):
+    """Feature metadata status paired with its LazyIncrement data.
+
+    This combines a pure Pydantic status model with the LazyIncrement object
+    needed for sample-level operations like generating previews.
+    """
+
+    status: FeatureMetadataStatus
+    lazy_increment: LazyIncrement | None
+
+
+def format_sample_previews(
+    lazy_increment: LazyIncrement,
+    added_count: int,
+    changed_count: int,
+    id_columns: Sequence[str] | None = None,
+    limit: int = 5,
+) -> list[str]:
+    """Format sample previews for added and changed samples.
+
+    Args:
+        lazy_increment: The LazyIncrement containing added/changed samples
+        added_count: Number of added samples (to avoid re-counting)
+        changed_count: Number of changed samples (to avoid re-counting)
+        id_columns: Columns to include in previews (defaults to ["sample_uid"])
+        limit: Maximum number of samples to preview per category
+
+    Returns:
+        List of formatted preview lines
+    """
+    lines: list[str] = []
+    cols = list(id_columns or ["sample_uid"])
+
+    if added_count > 0:
+        added_preview_df = (
+            lazy_increment.added.select(cols).head(limit).collect().to_polars()
+        )
+        if added_preview_df.height > 0:
+            preview_lines = [
+                ", ".join(f"{col}={row[col]}" for col in added_preview_df.columns)
+                for row in added_preview_df.to_dicts()
+            ]
+            lines.append("    Added samples: " + "; ".join(preview_lines))
+
+    if changed_count > 0:
+        changed_preview_df = (
+            lazy_increment.changed.select(cols).head(limit).collect().to_polars()
+        )
+        if changed_preview_df.height > 0:
+            preview_lines = [
+                ", ".join(f"{col}={row[col]}" for col in changed_preview_df.columns)
+                for row in changed_preview_df.to_dicts()
+            ]
+            lines.append("    Changed samples: " + "; ".join(preview_lines))
+
+    return lines
+
+
+def count_lazy_rows(lazy_frame: nw.LazyFrame[Any]) -> int:
+    """Return row count for a Narwhals LazyFrame.
+
+    Args:
+        lazy_frame: The LazyFrame to count rows from
+
+    Returns:
+        Number of rows in the LazyFrame
+    """
+    return lazy_frame.select(nw.len()).collect().to_polars()["len"].item()
+
+
+def get_feature_metadata_status(
+    feature_key: FeatureKey | type[BaseFeature],
```

Comment:
we have `metaxy.CoercibleToFeatureKey` and `ValidatedFeatureKeyTypeAdapter` - use them here (the base metadata store uses them a lot, look for examples there)